### PR TITLE
Add medicine count in dosage history when new medicine is added

### PR DESCRIPTION
**Pull request by astha-singh-dev**

### CODE REVIEW

File: src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java [hunk 1]
```diff
@@ -58,6 +65,13 @@ public MedicineResponseDto addMedicine(MedicineRequestDto medicineRequestDto, Lo
             throw new ResponseException(HttpStatus.BAD_REQUEST, ErrorMessages.SAME_MEDICINE_IS_ALREADY_EXIST);
         }
         Medicine savedMedicine = medicineRepository.save(medicine);
+        DosageHistory dosageHistory = new DosageHistory();
+        dosageHistory.setMedicineId(savedMedicine.getId());
+        dosageHistory.setMedicine(savedMedicine);
+        dosageHistory.setDosage(savedMedicine.getCurrentDosage());
+        dosageHistory.setType(DosageType.REFILL);
+        dosageHistory.setUserId(userId);
```
Comment (ShivamNagpal) [inline src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java:73]: Move this to a transformer

File: src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java [hunk 2]
```diff
@@ -164,4 +178,15 @@ public List<String> searchMedicine(Long userId, String medicineName) {
         return medicine;
     }
 
+    @Override
+    public void setReminder(Long medicineId, Integer dosageCount, Long userId) {
+        Optional<Medicine> optionalMedicine = medicineRepository.findByIdAndUserIdAndDeleted(medicineId, userId, false);
+        if (optionalMedicine.isEmpty()) {
+            throw new ResponseException(HttpStatus.BAD_REQUEST, ErrorMessages.MEDICINE_NOT_EXIST);
```
Comment (ShivamNagpal) [inline src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java:185]: Use NOT_FOUND instead of BAD_REQUEST

Comment (astha-singh-2445) [reply src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java]: resolved

File: src/main/java/com/singh/astha/medicinereminder/dtos/transformers/DosageDtoTransformer.java [hunk 3]
```diff
@@ -5,11 +5,22 @@
 import com.singh.astha.medicinereminder.dtos.ResponseDto.MedicineResponseDto;
 import com.singh.astha.medicinereminder.enums.DosageType;
 import com.singh.astha.medicinereminder.models.DosageHistory;
+import com.singh.astha.medicinereminder.models.Medicine;
 import org.springframework.stereotype.Component;
 
 @Component
 public class DosageDtoTransformer {
 
+    public static DosageHistory setDosageHistory(Long userId, Medicine savedMedicine) {
```
Comment (ShivamNagpal) [inline src/main/java/com/singh/astha/medicinereminder/dtos/transformers/DosageDtoTransformer.java:14]: This method should not be static.

File: src/main/java/com/singh/astha/medicinereminder/dtos/transformers/DosageDtoTransformer.java [hunk 4]
```diff
@@ -5,11 +5,22 @@
...
+        DosageHistory dosageHistory = new DosageHistory();
+        dosageHistory.setMedicineId(savedMedicine.getId());
+        dosageHistory.setMedicine(savedMedicine);
+        dosageHistory.setDosage(savedMedicine.getCurrentDosage());
+        dosageHistory.setType(DosageType.REFILL);
+        dosageHistory.setUserId(userId);
```
Comment (ShivamNagpal) [inline src/main/java/com/singh/astha/medicinereminder/dtos/transformers/DosageDtoTransformer.java:20]: This method shouldn't be a transformer.

Create a private method in the same call and call it as initialiseDosageHistory

File: src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java [hunk 5]
```diff
@@ -38,16 +42,19 @@ public class MedicineServiceImpl implements MedicineService {
...
     private final CategoryDtoTransformer categoryDtoTransformer;
 
+    private final DosageHistoryRepository dosageHistoryRepository;
+
     @Autowired
     public MedicineServiceImpl(MedicineRepository medicineRepository, MedicineDtoTransformer medicineDtoTransformer,
                                CategoryRepository categoryRepository,
                                MedicineCategoryRepository medicineCategoryRepository,
-                               CategoryDtoTransformer categoryDtoTransformer) {
+                               CategoryDtoTransformer categoryDtoTransformer, DosageService dosageService, DosageHistoryRepository dosageHistoryRepository) {
```
Comment (ShivamNagpal) [inline src/main/java/com/singh/astha/medicinereminder/services/impl/MedicineServiceImpl.java:51]: I don't see any usage of `dosageService` here.

Also don't wire repository of some other domain here directtly. Instead create a DosageHelper and create a helper method for your  usecase.

**Review by ShivamNagpal — APPROVED** [review]

**Pull request merged**